src/main/java/io/cdap/plugin/batch/source/BSONObjectToRecordTransformer.java

```diff
@@ -0,0 +1,201 @@
+    this.schema = schema;
+  }
+
+  /**
+   * Transforms given {@link BSONObject} to {@link StructuredRecord}.
+   *
+   * @param bsonObject BSON object to be transformed.
+   * @return {@link StructuredRecord} that corresponds to the given {@link BSONObject}.
+   */
+  public StructuredRecord transform(BSONObject bsonObject) {
+    return convertRecord(null, bsonObject, schema);
+  }
+
+  private StructuredRecord convertRecord(@Nullable String fieldName, BSONObject object, Schema schema) {
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    List<Schema.Field> fields = Objects.requireNonNull(schema.getFields(), "Schema fields cannot be empty");
+    for (Schema.Field field : fields) {
+      // Use full field name for nested records to construct meaningful errors messages.
+      // Full field names will be in the following format: 'record_field_name.nested_record_field_name'
+      String fullFieldName = fieldName != null ? String.format("%s.%s", fieldName, field.getName()) : field.getName();
+      Schema nonNullableSchema = field.getSchema().isNullable() ? field.getSchema().getNonNullable()
+        : field.getSchema();
+      Object value = extractValue(fullFieldName, object.get(field.getName()), nonNullableSchema);
+
+      Schema.LogicalType fieldLogicalType = nonNullableSchema.getLogicalType();
+      if (fieldLogicalType == null) {
+        // Set values of simple types
+        builder.set(field.getName(), value);
+        continue;
+      }
+
+      // Set values of logical types properly
+      switch (fieldLogicalType) {
+        case TIMESTAMP_MILLIS:
+        case TIMESTAMP_MICROS:
+          builder.setTimestamp(field.getName(), (ZonedDateTime) value);
+          break;
+        case DECIMAL:
+          builder.setDecimal(field.getName(), (BigDecimal) value);
+          break;
+        default:
+          throw new UnexpectedFormatException("Field type '" + fieldLogicalType + "' is not supported.");
+      }
+    }
+    return builder.build();
+  }
+
+  private Object extractValue(String fieldName, Object object, Schema schema) {
+    if (object == null || object instanceof BsonUndefined) {
+      return null;
+    }
+
+    Schema.LogicalType fieldLogicalType = schema.getLogicalType();
+    if (fieldLogicalType != null) {
+      switch (fieldLogicalType) {
+        case TIMESTAMP_MILLIS:
+        case TIMESTAMP_MICROS:
+          ensureTypeValid(fieldName, object, Date.class);
+          Date date = (Date) object;
+          return date.toInstant().atZone(ZoneId.ofOffset("UTC", ZoneOffset.UTC));
+        case DECIMAL:
+          ensureTypeValid(fieldName, object, Decimal128.class);
+          Decimal128 decimal = (Decimal128) object;
+          int scale = schema.getScale();
+          return decimal.bigDecimalValue().setScale(scale, BigDecimal.ROUND_HALF_EVEN);
+        default:
+          throw new UnexpectedFormatException("Field type '" + fieldLogicalType + "' is not supported.");
```

Comment:
error message should include which field is invalid.

Field 'X' is of unsupported type 'Y'

Reply:
Changed.